File: CodingInterviews/LinkList/GetIntersectionNode.java
package LinkList;

class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null)
            return null;
        ListNode pa = headA, pb = headB;
        while (pa != pb) {
            pa = pa.next;
            pb = pb.next;
            if (pa == null && pb == null)
                return null;
            if (pa == null)
                pa = headB;
            if (pb == null)
                pb = headA;
        }
        return pa;
    }
}

File: LeetCode/Explore/LinkedList/GetIntersectionNode.cpp
#include "main.h"

class Solution {
  public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if (headA == nullptr || headB == nullptr)
            return nullptr;
        auto pa = headA, pb = headB;
        while (pa != pb) {
            pa = pa->next;
            pb = pb->next;
            if (pa == nullptr && pb == nullptr)
                return nullptr;
            if (pa == nullptr)
                pa = headB;
            if (pb == nullptr)
                pb = headA;
        }
        return pa;
    }
};